File: lib/trends/youtube.ts
// YouTube Data API v3. Requires YOUTUBE_API_KEY.
// Strategy: keyword search filtered to videoCategoryId=20 (Gaming), order by viewCount,
// then fetch video stats. Returns top videos + aggregate view count.

import { wrap, fetchJson, fmtNum } from './_helpers'
import type { TrendQuery, TrendResult, TrendDatapoint } from './types'

type SearchResp = { items: Array<{ id: { videoId: string }; snippet: { title: string; channelTitle: string; publishedAt: string } }> }
type StatsResp = { items: Array<{ id: string; statistics: { viewCount?: string; likeCount?: string; commentCount?: string }; snippet: { title: string; channelTitle: string; publishedAt: string } }> }

export async function fetchYoutube(query: TrendQuery): Promise<TrendResult> {
  return wrap('youtube', query, async () => {
    const key = process.env.YOUTUBE_API_KEY
    if (!key) throw new Error('YOUTUBE_API_KEY not set')
    const keywords = (query.keywords ?? []).slice(0, 3)
    if (keywords.length === 0) throw new Error('no keywords')

    const q = keywords.join(' ')
    const regionCode = (query.region ?? 'US').toUpperCase()
    const searchUrl = `https://www.googleapis.com/youtube/v3/search?part=snippet&q=${encodeURIComponent(q)}&type=video&videoCategoryId=20&order=viewCount&maxResults=10&regionCode=${regionCode}&key=${key}`
    const search = await fetchJson<SearchResp>(searchUrl)
    const ids = (search.items ?? []).map(i => i.id.videoId).filter(Boolean)
    if (ids.length === 0) throw new Error('no videos')

    const statsUrl = `https://www.googleapis.com/youtube/v3/videos?part=statistics,snippet&id=${ids.join(',')}&key=${key}`
    const stats = await fetchJson<StatsResp>(statsUrl)
    const rows = (stats.items ?? []).map(v => ({
      title: v.snippet.title,
      channel: v.snippet.channelTitle,
      views: Number(v.statistics.viewCount ?? 0),
      likes: Number(v.statistics.likeCount ?? 0),
      published: v.snippet.publishedAt,
    }))
    rows.sort((a, b) => b.views - a.views)

    const totalViews = rows.reduce((s, r) => s + r.views, 0)
    const datapoints: TrendDatapoint[] = rows.slice(0, 10).map(r => ({
      label: r.title.slice(0, 40),
      value: r.views,
      meta: { channel: r.channel, likes: r.likes },
    }))
    const summary = `YouTube ${keywords.join(' / ')} 区域 ${regionCode}，Top 10 视频总播放 ${fmtNum(totalViews)}，第一名 ${fmtNum(rows[0].views)}。`
    const digest = [
      `# YouTube · ${keywords.join(' / ')} · ${regionCode}`,
      '',
      `**Top 10 视频总播放**：${fmtNum(totalViews)}`,
      '',
      ...rows.slice(0, 5).map((r, i) =>
        `${i + 1}. ${r.title.slice(0, 70)} (${r.channel}) · ${fmtNum(r.views)} 播放 · ${fmtNum(r.likes)} 赞`,
      ),
    ].join('\n')
    return { summary, digest, datapoints }
  })
}

File: server/stratsquad/trends/youtube.py
"""YouTube Data API v3. Needs YOUTUBE_API_KEY."""
from __future__ import annotations
import os
from urllib.parse import quote

from ..types import TrendQuery, TrendDatapoint, TrendResult
from .base import wrap, fetch_json, fmt_num


async def fetch_youtube(query: TrendQuery) -> TrendResult:
    async def body() -> dict:
        key = os.getenv("YOUTUBE_API_KEY")
        if not key:
            raise RuntimeError("YOUTUBE_API_KEY not set")
        keywords = (query.keywords or [])[:3]
        if not keywords:
            raise ValueError("no keywords")
        q = quote(" ".join(keywords))
        region = (query.region or "US").upper()

        search_url = (
            f"https://www.googleapis.com/youtube/v3/search?part=snippet&q={q}"
            f"&type=video&videoCategoryId=20&order=viewCount&maxResults=10"
            f"&regionCode={region}&key={key}"
        )
        search = await fetch_json(search_url)
        ids = [i["id"]["videoId"] for i in (search.get("items") or []) if i.get("id", {}).get("videoId")]
        if not ids:
            raise ValueError("no videos")
        stats_url = (
            f"https://www.googleapis.com/youtube/v3/videos?part=statistics,snippet"
            f"&id={','.join(ids)}&key={key}"
        )
        stats = await fetch_json(stats_url)
        rows = [
            {
                "title": v["snippet"]["title"],
                "channel": v["snippet"]["channelTitle"],
                "views": int(v.get("statistics", {}).get("viewCount", 0)),
                "likes": int(v.get("statistics", {}).get("likeCount", 0)),
            }
            for v in stats.get("items", [])
        ]
        rows.sort(key=lambda r: r["views"], reverse=True)
        total_views = sum(r["views"] for r in rows)
        datapoints = [
            TrendDatapoint(label=r["title"][:40], value=float(r["views"]), meta={"channel": r["channel"], "likes": r["likes"]})
            for r in rows[:10]
        ]
        summary = f"YouTube {' / '.join(keywords)} 区域 {region}，Top 10 视频总播放 {fmt_num(total_views)}，第一名 {fmt_num(rows[0]['views'])}。"
        digest_lines = [
            f"# YouTube · {' / '.join(keywords)} · {region}",
            "",
            f"**Top 10 视频总播放**：{fmt_num(total_views)}",
            "",
        ]
        for i, r in enumerate(rows[:5]):
            digest_lines.append(f"{i+1}. {r['title'][:70]} ({r['channel']}) · {fmt_num(r['views'])} 播放 · {fmt_num(r['likes'])} 赞")
        return {"summary": summary, "digest": "\n".join(digest_lines), "datapoints": datapoints}

    return await wrap("youtube", query, body)
